keps/sig-architecture/4330-compatibility-versions/README.md

```diff
@@ -0,0 +1,1179 @@
+- keep binary-version 1.31 while upgrading compat-version to 1.30 (stepwise upgrade of compatibility)
+- keep binary-version 1.31 while upgrading compat-version to 1.31 (stepwise upgrade of compatibility)
+
+Benefits to upgrading binary version independent of compatibility version:
+
+- A skip-version binary upgrade that transitions through compat versions has
+  higher probability of bugs in those intermediate binary versions already being
+  fixed.
+- During an upgrade, it is possible upgrade the new binary version while the
+  compatibility version is fixed (e.g. `binaryVersion: 1.28` -> `{binaryVersion: 1.29, compatVersion: 1.28}`).
+  This allow differences in the binary (bugs fixed or introduced, performance
+  changes, ...) to be introduced into a cluster and verified safe before
+  allowing access to the new APIs and features new version. Once the binary
+  version has been deamed safe, the compatibility version can then be upgraded.
+- Any upgrade system that successfully detects failures between upgrade steps
+  can report which upgrade step failed. This makes it easier to diagnose the
+  failures, because there are fewer possible causes of the failure. (There's a
+  huge difference between "A cluster upgrade failed" and "A cluster upgrade
+  failed when only the apiserver's binary version changed").
+- For upgrades where multiple failures occur, this increases the odds those
+  failures are split across steps. An upgrade system that is able to pause after
+  a step where failures are detected allow for failures at that step to be
+  addressed before proceeding to subsequent steps. These failures can be
+  addressed without the disruption and "noise" from failures in subsequent
+  steps.
+- A compatibility version rollback can be performed without changing binary
+  version.
+
+
+### Goals
+
+- Introduce the metadata necessary to configure features/APIs/storage-versions/CEL
+  features to match the behavior of an older Kubernetes release version
+- A Kubernetes binary with compatibility version set to N, will pass the
+  conformance and e2e tests from Kubernetes release version N.
+- A Kubernetes binary with compatibility version set to N does not enable any
+  changes (storage versions, CEL feature, features) that would prevent it
+  from being rolled back to N-1.
+- The most recent Kubernetes version supports compatibility version being set to
+  the full range of supported versions (N..N-3).
+
+### Non-Goals
+
+- Support compatibility version for Alpha features.  Alpha feature are not
+  designed to be upgradable, so we will not allow alpha features to be used in
+  conjunction with compatibility version.
+- Changes to Cluster API/kubeadm/KIND/minikube to absorb the compatibility versions
+  will be addressed separate from this KEP
+
+## Proposal
+
+### Component Flags
+
+Kubernetes components (apiservers, controller managers, schedulers) will offer a
```

Comment:
How about kubelet (or node components in general)?
We probably don't need it, but it would be userful to state it explicitly one side or the other.

Reply:
Good catch. Yes, consider the kubelet out of scope for this enhancement.  I'll make note in the KEP.

The rationale here is: a) kubelet already supports skip level upgrades and N-3 skew. b) there is a lot of complexity in node components that would have to be addressed specifically that it beyond what we'd like to take on here

Reply:
@logicalhan, does that match what you were thinking?

Today, skew requirements say the API server version must be >= the kubelet versions speaking to it. 

If node components will not make use of `--compatibility-version` info, then at minimum, that needs clarification to say the API server binary *and* compatibility version must be >= the kubelet versions speaking to it.

Reply:
oh yeah - this is what pushed me to this comment, but then I eventually forget to add it (since I was editing it multiple time and got distracted in the meantime)

yeah - we need to provide guarantees about compatibility version and kubelet version.

@jpbetz - please add

Reply:
Thanks, I've clarified.